[FuncSpec] Query SCCPSolver in more places

hazzlim opened the pull request:
When traversing the use-def chain of an Argument in a candidate specialization, also query the SCCPSolver to see if a Value is constant. This allows us to better estimate the codesize savings of a candidate in the presence of instructions that are a user of the argument we are estimating savings for which also use arguments that have been found constant by IPSCCP.

Similarly when estimating the dead basic blocks from branch and switch instructions which become constant, also query the SCCPSolver to see if a predecessor is unreachable.

hazzlim committed:
[FuncSpec] Query SCCPSolver in more places

When traversing the use-def chain of an Argument in a candidate
specialization, also query the SCCPSolver to see if a Value is constant.
This allows us to better estimate the codesize savings of a candidate
in the presence of instructions that are a user of the argument we are
estimating savings for which also use arguments that have been found
constant by IPSCCP.

Similarly when estimating the dead basic blocks from branch and switch
instructions which become constant, also query the SCCPSolver to see if
a predecessor is unreachable.

ChuanqiXu9 reviewed (approved):
LGTM generally.

david-arm commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -312,10 +317,10 @@ bool InstCostVisitor::discoverTransitivelyIncomingValues(
 
       // Disregard self-references and dead incoming values.
       if (auto *Inst = dyn_cast<Instruction>(V))
-        if (Inst == PN || DeadBlocks.contains(PN->getIncomingBlock(I)))
```
Would that suggest a bug then? Is there a way for the solver InstCostVisitor to reuse the same code that finds a basic block dead?

hazzlim replied:
It's maybe confusing that this function is named the same as the `SCCPSolver` member function, but this change is to make us dispatch to `InstCostVisitor::isBlockExecutable`, which checks both the `SCCPSolver` and DeadBlocks:
https://github.com/llvm/llvm-project/blob/2de3d00edc614cdce4b30b68be564a7a3123dbf9/llvm/include/llvm/Transforms/IPO/FunctionSpecialization.h#L179-L181

labrinea replied:
> Would that suggest a bug then?

No it's not a bug.

>  Is there a way for the solver and InstCostVisitor to reuse the same code that finds a basic block dead?

I am not sure I understand the question. The InstCostVisitor is a simplified version of the Solver for the pursposes of estimating the profitability of specialization. If the specialization is performed the Solver will end up matching the findings of InstCostVisitor.

labrinea replied:
> It's maybe confusing that this function is named the same as the SCCPSolver member function, but this change is to make us dispatch to InstCostVisitor::isBlockExecutable

Ah right, excuse me I had forgotten that. Then it makes sense. Can you use that function inside canEliminateSuccessor then?

hazzlim replied:
> Ah right, excuse me I had forgotten that. Then it makes sense. Can you use that function inside canEliminateSuccessor then?

Yep, will do this as well as updating things to be non-static/member functions, as per other comments.

hazzlim replied:
> Can you use that function inside canEliminateSuccessor then?

hazzlim committed:
Address review comments

- Make canEliminateSuccessor a non-static, const member function
- Make findConstantFor a non-static, const member function

hazzlim commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -91,12 +91,14 @@ static cl::opt<bool> SpecializeLiteralConstant(
 bool InstCostVisitor::canEliminateSuccessor(BasicBlock *BB, BasicBlock *Succ,
-                                         DenseSet<BasicBlock *> &DeadBlocks) {
+                                            DenseSet<BasicBlock *> &DeadBlocks,
+                                            const SCCPSolver &Solver) {
```
Good point - updated to be non-static (and const).

hazzlim commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -135,15 +137,18 @@ Cost InstCostVisitor::estimateBasicBlocks(
 }
 
-static Constant *findConstantFor(Value *V, ConstMap &KnownConstants) {
+static Constant *findConstantFor(Value *V, ConstMap &KnownConstants,
```
Good point - done.

hazzlim committed:
Add assert when inserting into DeadBlocks

hazzlim commented on llvm/lib/Transforms/IPO/FunctionSpecialization.cpp:
```diff
@@ -347,10 +352,10 @@ Constant *InstCostVisitor::visitPHINode(PHINode &I) {
 
     // Disregard self-references and dead incoming values.
     if (auto *Inst = dyn_cast<Instruction>(V))
-      if (Inst == &I || DeadBlocks.contains(I.getIncomingBlock(Idx)))
```
I've left this as is, as per discussion above. I have added an assert just before inserting into DeadBlocks that the Solver thinks the block is executable.

labrinea reviewed (approved):
Looks good, thanks!

hazzlim commented:
Thank you for reviewing! :)

FlameTop commented:
We are currently experiencing assertion failures after this commit landed. A reproducer is 

https://godbolt.org/z/oTjdex7Gh

which is failing with the head of trunk at this date.

davemgreen commented:
It looks like it doesn't like the metadata on the strict-fp call args. We should be able to skip that when looking at the operands.

davemgreen commented:
#124284 hopefully addresses this. Thanks.
